### PR TITLE
nixos/traefik: add environmentFiles option

### PR DESCRIPTION
**Pull request by 999eagle**
###### Description of changes

This adds a new option to the traefik module so environment variables can be provided through files and substituted into traefik's static configuration. This can be used for secrets as traefik by itself doesn't support reading configuration from file and environment variables.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-servers/traefik.nix [hunk 1]
```diff
@@ -48,6 +48,11 @@ let
     ''
   else
     cfg.staticConfigFile;
+
+  finalStaticConfigFile =
+    if cfg.environmentFiles == []
+    then staticConfigFile
+    else "/var/run/traefik/config.toml";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/traefik.nix:55]:
```suggestion
    else "/etc/traefik/config.toml";
```
or just /run, nixos is not using /var/run

File: nixos/modules/services/web-servers/traefik.nix [hunk 2]
```diff
@@ -139,8 +154,14 @@ in {
...
       serviceConfig = {
+        EnvironmentFile = cfg.environmentFiles;
+        ExecStartPre = lib.optional (cfg.environmentFiles != [])
+          (pkgs.writeShellScript "pre-start" ''
+            umask 077
+            ${pkgs.envsubst}/bin/envsubst -i "${staticConfigFile}" > "${finalStaticConfigFile}"
+          '');
         ExecStart =
-          "${cfg.package}/bin/traefik --configfile=${staticConfigFile}";
+          "${cfg.package}/bin/traefik --configfile=${finalStaticConfigFile}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/traefik.nix:164]:
```suggestion
        ExecStart = "${cfg.package}/bin/traefik --configfile=${finalStaticConfigFile}";
```

File: nixos/modules/services/web-servers/traefik.nix (hunk 1 above)
Comment (999eagle) [reply nixos/modules/services/web-servers/traefik.nix]: Changed to `/run` so secrets from environment variables aren't leaked to `/etc`.

File: nixos/modules/services/web-servers/traefik.nix (hunk 2 above)
Comment (999eagle) [reply nixos/modules/services/web-servers/traefik.nix]: Fixed.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by 999eagle** [comment]
@SuperSandro2000 is there anything still left to do or can this be merged?

**Pull request merged**